--- a/OpenSSL/crypto/x509ext.c
+++ b/OpenSSL/crypto/x509ext.c
@@ -13,6 +13,23 @@
 #define crypto_MODULE
 #include "crypto.h"
 
+static char crypto_X509Extension_set_critical_doc[] = "\n\
+Returns the critical field of the X509Extension\n\
+";
+
+static PyObject *
+crypto_X509Extension_set_critical(crypto_X509ExtensionObj *self, PyObject *args)
+{
+    int critical;
+    if (!PyArg_ParseTuple(args, "i:get_critical", &critical))
+        return NULL;
+
+    X509_EXTENSION_set_critical(self->x509_extension, critical);
+    Py_INCREF(Py_None);
+    return Py_None;
+}
+
+
 static char crypto_X509Extension_get_critical_doc[] = "\n\
 Returns the critical field of the X509Extension\n\
 \n\
@@ -105,6 +122,7 @@
 static PyMethodDef crypto_X509Extension_methods[] =
 {
     ADD_METHOD(get_critical),
+    ADD_METHOD(set_critical),
     ADD_METHOD(get_short_name),
     ADD_METHOD(get_data),
     ADD_METHOD(get_oid),
